### PR TITLE
Improvements to Label's layout options

### PR DESCRIPTION
**Pull request by Geometror**
This PR aims to improve the layout options for the Label control. 
Implements godotengine/godot-proposals#2916 and godotengine/godot-proposals#2931
(This needed to be done at once since the additions are interdependent)
If these additions/changes get approved, it makes sense to add them (partly/fully) to TextLine and TextParagraph in a subsequent PR (or this?).
grafik

- Added options to trim the text in case it overruns (exceeds the node's bounding box)
    - Trim characters:
Animation1

    - Trim words:
Animation2

    - Ellipsis trimming:
Animation3

    - Ellipsis word trimming:
Animation4

- Added more autowrap modes
    - Arbitrary:
Animation5
    - Word (Smart):
Animation7
- Slight change to line breaking logic, which ignores single trailing spaces

New properties:
grafik
grafik

The only thing missing is proper RTL text support (the ellipsis string should be placed on the left then), but that will be added once I figured out how to do that properly. Also, I am not sure if I implemented it correctly or at the right place in the codebase but after digging through the code it seemed like the only option without changing huge parts of the code.

Any testing/feedback would be appreciated :)

### CODE REVIEW

File: modules/text_server_adv/text_server_adv.cpp [hunk 1]
```diff
@@ -1658,6 +1658,123 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
...
+	int ellipsis_pos = -1;
+	if (enforce_ellipsis) {
+		clip_pos = 0;
+		ellipsis_pos = 0;
+	}
+
+	int last_valid_cut = 0;
+	bool found = false;
+
+	for (int i = sd_size - 1; i >= 0; --i) {
```
Comment (bruvzg) [inline modules/text_server_adv/text_server_adv.cpp:1708]: This should run from the other end (and add ellipsis on the left side) if detected text direction (`sd->para_direction`) is RTL.

Screenshot 2021-07-02 at 19 23 47

File: scene/gui/label.cpp [hunk 2]
```diff
@@ -370,13 +420,13 @@ Size2 Label::get_minimum_size() const {
...
-		return Size2(1, clip ? 1 : min_size.height) + min_style;
+	if (autowrap_mode != AUTOWRAP_OFF) {
+		return Size2(1, (clip || overrun_behavior != OVERRUN_NO_TRIMMING) ? 1 : min_size.height) + min_style;
 	} else {
-		if (clip) {
+		if (clip || overrun_behavior != OVERRUN_NO_TRIMMING) {
 			min_size.width = 1;
 		}
-
+		auto res = min_size + min_style;
```
Comment (bruvzg) [inline scene/gui/label.cpp:429]: Using `auto` is usually against Godot code style.

File: servers/text_server.h [hunk 3]
```diff
@@ -66,8 +66,15 @@ class TextServer : public Object {
 		BREAK_NONE = 0,
 		BREAK_MANDATORY = 1 << 4,
 		BREAK_WORD_BOUND = 1 << 5,
-		BREAK_GRAPHEME_BOUND = 1 << 6
-		//RESERVED = 1 << 7
+		BREAK_GRAPHEME_BOUND = 1 << 6,
+		BREAK_WORD_BOUND_ADAPTIVE = 1 << 5 | 1 << 7
+	};
+
+	enum TextOverrunFlag {
```
Comment (bruvzg) [inline servers/text_server.h:73]: Missing `VARIANT_ENUM_CAST(TextServer::TextOverrunFlag);` at the end of file and `BIND_ENUM_CONSTANT(...)`s in the text server `_bind_methods()`.

File: modules/text_server_adv/text_server_adv.cpp [hunk 4]
```diff
@@ -1658,6 +1658,123 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
...
+			if (cut_per_word && (ellipsis_pos > 0)) {
+				TextServer::Glyph gl;
+				gl.start = sd_glyphs[ellipsis_pos].start;
+				gl.end = sd_glyphs[ellipsis_pos].end;
+				gl.count = 1;
+				gl.advance = whitespace_adv.x;
+				gl.index = whitespace_gl_idx;
+				gl.font_rid = last_gl_font_rid;
+				gl.font_size = last_gl_font_size;
+				gl.flags = GRAPHEME_IS_SPACE | GRAPHEME_IS_BREAK_SOFT;
```
Comment (bruvzg) [inline modules/text_server_adv/text_server_adv.cpp:1750]: The added glyphs probably should have `GRAPHEME_IS_VIRTUAL` bit set (to be non-selectable and skipped by caret movement), irrelevant for the label, but good for consistency.

File: scene/gui/label.cpp (hunk 2 above)
Comment (Geometror) [reply scene/gui/label.cpp]: Oops, that line was for debugging and I forgot to remove it. Fixed.

File: servers/text_server.h (hunk 3 above)
Comment (Geometror) [reply servers/text_server.h:73]: Fixed.

File: modules/text_server_adv/text_server_adv.cpp (hunk 4 above)
Comment (Geometror) [reply modules/text_server_adv/text_server_adv.cpp]: Done.

**Comment by Geometror** [comment]
Added RTL support and did some optimizations/refactoring
grafik

File: modules/text_server_adv/text_server_adv.cpp (hunk 1 above)
Comment (Geometror) [reply modules/text_server_adv/text_server_adv.cpp]: Implemented.

**Review by bruvzg — APPROVED** [review]

File: modules/text_server_adv/text_server_adv.cpp [hunk 5]
```diff
@@ -1658,6 +1658,161 @@ float TextServerAdvanced::shaped_text_tab_align(RID p_shaped, const Vector<float
...
+				added_glyphs++;
+			}
+		}
+
+		// Cut the remaining glyphs off.
+		if (!is_rtl) {
+			sd->glyphs.resize(trim_pos + added_glyphs);
+		} else {
+			for (int ridx = 0; ridx <= trim_pos - added_glyphs; ridx++) {
+				sd->glyphs.remove(0);
```
[the remark this answers is not in the document]
Comment (Geometror) [reply modules/text_server_adv/text_server_adv.cpp]: Done. I also had to fix another bug introduced by former optimizations in the code for handling RTL text.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**